components/raftstore-v2/src/worker/tablet_gc.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    fmt::{self, Display, Formatter},
+    path::{Path, PathBuf},
+    time::Duration,
+};
+
+use collections::HashMap;
+use engine_traits::{DeleteStrategy, KvEngine, Range, TabletContext, TabletRegistry};
+use slog::{warn, Logger};
+use tikv_util::worker::{Runnable, RunnableWithTimer};
+
+pub enum Task<EK> {
+    Trim {
+        tablet: EK,
+        start_key: Box<[u8]>,
+        end_key: Box<[u8]>,
+    },
+    PrepareDestroy {
+        region_id: u64,
+        tablet: EK,
+        wait_for_persisted: u64,
+    },
+    Destroy {
+        region_id: u64,
+        persisted_index: u64,
+    },
+}
+
+impl<EK> Display for Task<EK> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // TODO
+        write!(f, "TabletGcTask")
+    }
+}
+
+pub struct Runner<EK: KvEngine> {
+    tablet_registry: TabletRegistry<EK>,
+    logger: Logger,
+
+    pending_trim_tasks: Vec<InternalTrimTask<EK>>,
+    // region_id -> [(tablet_path, wait_for_persisted)].
+    waiting_destroy_tasks: HashMap<u64, Vec<(PathBuf, u64)>>,
+    pending_destroy_tasks: Vec<PathBuf>,
+}
+
+struct InternalTrimTask<EK> {
+    tablet: EK,
+    start_key: Box<[u8]>,
+    end_key: Box<[u8]>,
+    arrival_seqno: u64,
+}
+
+impl<EK: KvEngine> Runner<EK> {
+    pub fn new(tablet_registry: TabletRegistry<EK>, logger: Logger) -> Self {
+        Self {
+            tablet_registry,
+            logger,
+            pending_trim_tasks: Vec::new(),
+            waiting_destroy_tasks: HashMap::default(),
+            pending_destroy_tasks: Vec::new(),
+        }
+    }
+
+    /// Returns true if task is consumed. Failure is considered consumed.
+    fn process_trim_task(task: &InternalTrimTask<EK>) -> bool {
+        if task
```

Comment:
The check is redundant. One point of checkpoint is to delete all stale keys immediately.